core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1151,16 +1155,20 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     }
   }
 
-  case class TopicDeletionStopReplicaResponseReceived(stopReplicaResponseObj: AbstractResponse, replicaId: Int) extends ControllerEvent {
+  case class StopAndDeleteReplicaResponseReceived(stopReplicaResponseObj: AbstractResponse, replicaId: Int) extends ControllerEvent {
 
     def state = ControllerState.TopicDeletion
 
     override def process(): Unit = {
-      import JavaConverters._
       if (!isActive) return
       val stopReplicaResponse = stopReplicaResponseObj.asInstanceOf[StopReplicaResponse]
-      debug(s"Delete topic callback invoked for $stopReplicaResponse")
+
+      // Here we are only handling partitions which are part of topic deletion
       val responseMap = stopReplicaResponse.responses.asScala
+        .filterKeys(tp => controllerContext.isTopicQueuedUpForDeletion(tp.topic))
```

Comment:
We probably also want to check that the topic is eligible for deletion.

Reply:
I have a little trouble following the state machine. I think we will transition to `ReplicaDeletionIneligible` if either we get a `StopReplica` response with an error or if the replica fails during deletion. I think it would be possible for a pending `StopReplica` response to be received after the broker had been marked offline, so I think the suggestion makes sense. Just to be clear though, this seems to be an existing problem?

Reply:
I was concerned about the case when a topic is queued for deletion behind a pending partition reassignment. In that case, the pending reassignment could send StopReplicaRequest with deletePartition=true. In the current logic, there is no callback in that case. So, the replicas won't be transitioned to ReplicaDeletionIneligible or ReplicaDeletionSuccessful until the topic is eligible for deletion. I was just trying to preserve that logic now that we add a callback as long as deletePartition=true.

Reply:
Got it. That makes sense.

Reply:
@junrao Unfortunately, we have to skip the topic ineligible check in order to continue to `completeReplicaDeletion` while a replica is offline. Some test cases are failing because of this. I think we should be installing a callback anyway on reassignment so that we can retry, but we can leave this for a future patch. For now, I will see if I can find a nice way to preserve existing logic.